source/app/models/url.rb

```diff
@@ -0,0 +1,53 @@
+class Url < ActiveRecord::Base
+  require 'net/http'
+  
+  before_save :generate_shortened_url
+
+  validates :address, presence: true
+  validates :address, format: { with: /\Ahttps?:\/\/.*/, message: "must begin with 'http://' or 'https://'" }
+  validates :address, uniqueness: true
+  validate :url_is_correct_format, :url_responds_to_http_request
+
+  def url_is_correct_format
+  	unless address =~ /\A#{URI.regexp}\z/
+  	  errors.add(:address, "is not a valid url format")
+  	end
+  end
+
+  def url_responds_to_http_request
+  	url = URI.parse(address)
+  	begin
+  	  response = Net::HTTP.get_response(url)
+
+  	  if response.code == "404"
+  	    errors.add(:address, "is not accessible via HTTP(S)")
+  	  end
+  	rescue
+  	  errors.add(:address, "is not accessible via HTTP(S)")
+  	end
+  end
+
+  def generate_shortened_url
+  	url = URI.parse(address)
+  	if url.host.present?
```

Comment:
If you can move this check for host presence into a separate validator, then you can avoid having all of this code be nested inside an if-block.

Reply:
I redid the method per Jay's recommendation of random string generation and let the validators worry about URL validity